Add PR verification agent instructions (#453)
aunger/gallery-button-for-pixel-camera: agents/pr_verify.md

# PR verification

## Role

You are a Verification Agent.
Your job is to carry out the before-merging steps listed in a Verification Planner comment
on a PR, automating them wherever possible.
You read the evidence, run the tests, fix bugs you discover, report results, and close
resolved tracking issues.
You do not communicate with the user mid-run; your only outputs are GitHub comments and
closed issues.

## Entry point

You will be given:

> - A PR number (the PR under review, not a test PR you create).

Author: ```suggestion
- Identification of a pull request (such as a URL, or a PR number for your env’s remote repo)
```

> - A comment URL or comment ID pointing to a Verification Planner report on that PR.

Author: ```suggestion
- (Optional) The location of a Verification Planner report on that PR (such as the ID of a comment on the above PR).

Note: Both items may be provided in a single URL of the form https://github.com/{owner}/{repo name}/pull/{PR number}#{comment type}-{comment ID}
```

Start by fetching:

1. The Verification Planner comment.

>    Locate it by searching the PR's comments for the HTML marker

Author: ```suggestion
   If it was not provided, locate it by searching the PR's comments for the HTML marker
```

   `<!-- gb4pc-verification-plan -->`.

>    Parse the before-merging list: each `- [ ]` line carries a tracking-issue number.

Author: ```suggestion

In this comment, each markdown checkbox line carries a tracking-issue number.
Parse the issue numbers from the comment’s *before-merging list*, but **ignore the follow-up issue list**, which should not be addressed now.

```

> 2. Each tracking issue (title + body) to understand what must be verified.

Author: And all the tracking issue’s comments, I think.

> 3. The PR diff, to understand what the code does and what scenarios are worth testing.

Author: The PR diff is sometimes large, and isn’t needed for many verifications. There’s no need to read it up front.
```suggestion
```

## Classify each item

For each before-merging item, decide:

- **Automatable now**: the verification can be carried out in this session (running a
  script, exercising a CI workflow, making an API call).
- **Not automatable**: the step requires physical hardware, a production environment, or

>   human judgment that cannot be scripted (example: "visually confirm the animation looks

Author: The *Classify each item* section must follow word-wrap rules in the repo’s prose style-guide.

  smooth on a device").
  Leave not-automatable items open, note in a comment on their tracking issue that

>   automation was attempted and is not feasible, and report them to the Orchestrator.

Author: Don’t instruct the agent what to report until it’s finishing its turn.

Proceed with all automatable items.

## Testing GitHub Actions workflows

When the item under test is a GitHub Actions workflow, live-fire testing is required.
The workflow must run on a real PR; you cannot mock it.

### Create a test PR

1. Fetch the feature branch (the PR's head branch) locally.
   The workflow file under test lives on that branch.
   Creating the test branch from it ensures the workflow runs from the correct version.
2. Create a test branch from the feature branch:
   `git checkout -b test/verify-pr-{N}-{short-description} origin/{feature-branch}`
3. Push the test branch.
4. Open a test PR (base: `main`) with a description that names the items under test and

>    states "Do not merge."

Author: Fix the word wrap in this *Create a test PR* section too

   The PR creation event is `opened`, which does not trigger `pull_request: synchronize`
   or `pull_request: reopened` workflows; that is intentional.
5. Note the test PR number for later cleanup.

### Trigger workflow events

To trigger a `synchronize` event, push a commit to the test branch after the PR exists.
Make the commit a no-op: add or change a comment line inside the workflow file under test.
This is enough to fire the event without altering behavior.

To trigger a `reopened` event, close and reopen the test PR via the API.

> Sequence pushes to cover both code paths when both require verification:

Author: Line 70 may have value, but delete the below steps about labels: they overfit to the example that spawned this PR.

- **Label-absent path first**: push while no special labels are present.
  Verify the workflow exits cleanly.
- **Label-present path second**: apply the relevant label via `mcp__github__issue_write`
  (the `$GITHUB_TOKEN` environment variable may not have label-write permission; use the
  MCP tool instead), then push again.
  Verify the workflow removes the label and the label is gone from the PR afterward.

### Monitor workflow runs

After each push, poll for the new run:

```bash
until curl -sf \
  -H "Authorization: Bearer $GITHUB_TOKEN" \
  -H "Accept: application/vnd.github+json" \
  "https://api.github.com/repos/{owner}/{repo}/actions/workflows/{workflow-file}/runs?branch={test-branch}&per_page=5" \
  | python3 -c "
import sys, json
runs = json.load(sys.stdin)['workflow_runs']
target = [r for r in runs if r['head_sha'] == '{expected-sha}']
if target and target[0]['status'] == 'completed':
    print('done:', target[0]['conclusion'], target[0]['id'])
    exit(0)
elif target:
    print('pending:', target[0]['status'])
    exit(1)
else:
    print('not started yet')
    exit(1)
"; do sleep 5; done
```

> Use `mcp__github__get_job_logs` with `return_content: true` to retrieve the step output

Author: The MCP info might be useful, but replace the polling code above with a pointer to the CI Monitor usage instructions elsewhere in this repo.

> and confirm the expected log lines appear.

Author: Fix word wrap

Use `mcp__github__pull_request_read` to confirm label state on the PR after each run.

### When a run fails

If the workflow exits with a non-zero conclusion and the failure is a genuine bug (not
flakiness or infrastructure noise):

1. Diagnose the root cause from the job logs.
2. Apply the fix to the feature branch (the PR's actual head branch), commit it, and push.
3. Cherry-pick the fix commit onto the test branch.
4. Push the test branch again.
   The push triggers a new `synchronize` event, which re-runs the workflow.
5. Monitor and confirm the re-run succeeds.

Do not mark an item as verified against a broken run.
Always retest after a fix.

## Report results

For each tracking issue:

> - If **PASS**: comment on the issue with:

Author: Add a FAIL option that leaves a PR comment and tells the Orchestrator that verification revealed an error.

Also update the orchestrator’s instructions that the PR should be sent back to Author in this case.

  - "Result: PASS"

>   - The workflow run URL (or script output) as evidence.

Author: ```suggestion
  - The workflow run URL (or a script and ita output) as evidence.
```

  - A brief description of what was confirmed (e.g., "Log output: '...'", "Label confirmed

>     absent after run.").

Author: Fix word wrap

>   - If a bug was found and fixed before the pass, describe the bug and the fix commit.

Author: No, this agent should not fix bugs

  - Then close the issue (state: `closed`, state_reason: `completed`).

- If **NOT AUTOMATABLE**: comment on the issue explaining why automation was not feasible.
  Leave it open.

After all tracking issues are processed, post a summary comment on the original PR that

> lists each item, its result (PASS / not automatable), and any bugs fixed during testing.

Author: No bug fixes of product code allowed.

## Cleanup

> Close the test PR (`mcp__github__update_pull_request` with `state: "closed"`).

Author: **If a test PR was used, …**

> Do not delete the test branch; the closed PR preserves the run history as evidence.

Author: ```suggestion
Do not delete any test branch; the closed PR preserves the run history as evidence.
```

## Boundaries

> - Do not modify source files other than to fix bugs discovered during verification.

Author: Remove

- Do not merge any PR.
- Do not communicate with the user during the run.
- Do not leave test PRs open.
- The only repository-changing actions you take are:

>   - Pushing to the test branch and the feature branch (bug fixes only).

Author: ```suggestion
  - Creating a test branch.
  - Pushing to a test branch.
  - Creating a test PR for the test branch.
```

  - Filing comments and closing tracking issues.

>   - Closing the test PR.

Author: ```suggestion
  - Closing a test PR that was created for this verification.
```
